Step date locale

EmilHvitfeldt opened the pull request:
This PR aims to close #1000 by adding a `locale` argument to `step_date()`. This accomplished 2 things; First it removed the buggy behavior shown in #1000 by using the same transformation at `bake()` time as was used at `prep()` time. Secondly it allows you to set the locale up front as an argument so you can specify the locale without setting environment variables. 

``` r
library(recipes)

Sys.setlocale("LC_TIME", 'fr_FR.UTF-8')
#> [1] "fr_FR.UTF-8"

examples <- tibble(someday = lubridate::ymd("2000-12-20") + lubridate::days(0:40))
rec <- recipe(~ someday, examples) %>%
  step_date(all_predictors()) %>%
  prep() 

rec %>%
  bake(new_data = examples)
#> # A tibble: 41 × 4
#>    someday    someday_dow someday_month someday_year
#>    <date>     <fct>       <fct>                <dbl>
#>  1 2000-12-20 Mer         déc                   2000
#>  2 2000-12-21 Jeu         déc                   2000
#>  3 2000-12-22 Ven         déc                   2000
#>  4 2000-12-23 Sam         déc                   2000
#>  5 2000-12-24 Dim         déc                   2000
#>  6 2000-12-25 Lun         déc                   2000
#>  7 2000-12-26 Mar         déc                   2000
#>  8 2000-12-27 Mer         déc                   2000
#>  9 2000-12-28 Jeu         déc                   2000
#> 10 2000-12-29 Ven         déc                   2000
#> # … with 31 more rows

Sys.setlocale("LC_TIME", 'en_GB.UTF-8')
#> [1] "en_GB.UTF-8"

rec %>%
  bake(new_data = examples)
#> # A tibble: 41 × 4
#>    someday    someday_dow someday_month someday_year
#>    <date>     <fct>       <fct>                <dbl>
#>  1 2000-12-20 Mer         déc                   2000
#>  2 2000-12-21 Jeu         déc                   2000
#>  3 2000-12-22 Ven         déc                   2000
#>  4 2000-12-23 Sam         déc                   2000
#>  5 2000-12-24 Dim         déc                   2000
#>  6 2000-12-25 Lun         déc                   2000
#>  7 2000-12-26 Mar         déc                   2000
#>  8 2000-12-27 Mer         déc                   2000
#>  9 2000-12-28 Jeu         déc                   2000
#> 10 2000-12-29 Ven         déc                   2000
#> # … with 31 more rows
```

<sup>Created on 2022-06-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>



``` r
library(recipes)

Sys.setlocale("LC_TIME", 'fr_FR.UTF-8')
#> [1] "fr_FR.UTF-8"

examples <- tibble(someday = lubridate::ymd("2000-12-20") + lubridate::days(0:40))
rec <- recipe(~ someday, examples) %>%
  step_date(all_predictors(), locale = 'en_GB.UTF-8') %>%
  prep() 

rec %>%
  bake(new_data = examples)
#> # A tibble: 41 × 4
#>    someday    someday_dow someday_month someday_year
#>    <date>     <fct>       <fct>                <dbl>
#>  1 2000-12-20 Wed         Dec                   2000
#>  2 2000-12-21 Thu         Dec                   2000
#>  3 2000-12-22 Fri         Dec                   2000
#>  4 2000-12-23 Sat         Dec                   2000
#>  5 2000-12-24 Sun         Dec                   2000
#>  6 2000-12-25 Mon         Dec                   2000
#>  7 2000-12-26 Tue         Dec                   2000
#>  8 2000-12-27 Wed         Dec                   2000
#>  9 2000-12-28 Thu         Dec                   2000
#> 10 2000-12-29 Fri         Dec                   2000
#> # … with 31 more rows
```

<sup>Created on 2022-06-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

juliasilge commented on R/date.R:
```diff
@@ -27,6 +27,9 @@
 #'  FALSE`.
 #' @param ordinal A logical: should factors be ordered? Only
 #'  available for features `month` or `dow`.
+#' @param locale locale to be used for `month` and `dow`, see [locales].
```
```suggestion
#' @param locale Locale to be used for `month` and `dow`, see [locales].
```

juliasilge reviewed (commented):
We'll need to handle old recipes that don't have the locale at `bake()` time, right? We have tests for this for `keep_original_cols` so we can do this similarly. Or am I misunderstanding maybe?

EmilHvitfeldt commented:
You are not miss-understanding, I have fixed this and added another test to make sure it works.

So this PR will not fix recipes that were created using locale A with version 0.2.0 and used in version >0.2.0. Since it is impossible(?) to detect what locale was set during prepping. But it will no longer "break", and it will try the system local set.

If you knew what locale was used to prep the old recipe you could repair it using

```r

rec <- readRDS("awesome-rec.rds")

rec$steps[[1]]$locale <- 'fr_FR.UTF-8'
```

EmilHvitfeldt commented:
This PR is not all the way ready, I need to properly use [test fixtures](https://testthat.r-lib.org/articles/test-fixtures.html)

juliasilge reviewed (approved):
This looks great! I think that `object$locale %||% Sys.getlocale("LC_TIME")` is the best thing to do at `bake()` time. 👍
